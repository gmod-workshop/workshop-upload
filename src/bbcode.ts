import { Converter } from "@gmod-workshop/steamdown";

/**
 * Convert a markdown string to BBCode
 * @param markdown The markdown to convert
 * @returns The converted BBCode
 */
export async function convert(markdown: string): Promise<string> {
    const converter = new Converter();

    const bbcode = await converter.convert(markdown);

    return bbcode
        .replace(/https:\/\/github\.com\/([^\/]+\/[^\/]+)\/(issues|pull)\/(\d+)/g, '[url=https://github.com/$1/$2/$3]#$3[/url]')
        .trim();
}
