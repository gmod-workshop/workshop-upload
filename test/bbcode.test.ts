import { describe, test, expect } from "vitest";
import { convert } from "../src/bbcode.js";

describe("BBCode", () => {
    test("Converts markdown to BBCode", async () => {
        const markdown = `# Test

This is a test.

## Subheading

This is a subheading.

### Subsubheading

This is a subsubheading.

#### Subsubsubheading

This is a subsubsubheading.

##### Subsubsubsubheading

This is a subsubsubsubheading.

###### Subsubsubsubsubheading

This is a subsubsubsubsubheading.

- This is a list item.
- This is another list item.
- This is a third list item.
`;

        const bbcode = await convert(markdown);

        expect(bbcode).toMatchSnapshot();
    });

});
