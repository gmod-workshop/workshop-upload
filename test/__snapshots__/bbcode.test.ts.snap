// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BBCode > Converts markdown to BBCode 1`] = `
"[h1]Test[/h1]

This is a test.

[h2]Subheading[/h2]

This is a subheading.

[h3]Subsubheading[/h3]

This is a subsubheading.

[h4]Subsubsubheading[/h4]

This is a subsubsubheading.

[h5]Subsubsubsubheading[/h5]

This is a subsubsubsubheading.

[h6]Subsubsubsubsubheading[/h6]

This is a subsubsubsubsubheading.

[list]
	[*] This is a list item.
	[*] This is another list item.
	[*] This is a third list item.
[/list]"
`;
